chain/chain/src/chain.rs

```diff
@@ -1460,47 +1464,82 @@ impl Chain {
                 return Err(e);
             }
         };
+        let (apply_chunk_work, block_preprocess_info) = preprocess_res;
+        let block = block.into_inner();
+        let block_hash = *block.hash();
+        self.processing_blocks
+            .add_preprocessed_block(block, block_preprocess_info)
+            .map_err(|_err| ErrorKind::TooManyProcessingBlocks)?;
 
         // 2) apply chunks, this is where the transactions and receipts are processed. At this step,
         //    there still no change to ChainStore.
-        let (apply_chunk_work, block_preprocess_info) = preprocess_res;
-        let apply_results = do_apply_chunks(apply_chunk_work);
+        self.async_apply_chunks(block_hash, apply_chunk_work);
 
-        // 3) finally, store the block on chain. Here we write all the necessary changes in chain_update,
-        //    which will be committed to storage all at once.
-        let mut chain_update = self.chain_update();
-        let new_head =
-            chain_update.postprocess_block(me, &block, block_preprocess_info, apply_results)?;
-        chain_update.chain_store_update.save_block_height_processed(block_height);
-        chain_update.commit()?;
+        // TODO: this will be moved out of process_block once we make do_apply_chunks async
+        self.try_postprocess_block(me, block_accepted)
+    }
 
-        self.pending_states_to_patch = None;
+    fn async_apply_chunks(
+        &self,
+        block_hash: CryptoHash,
+        work: Vec<Box<dyn FnOnce() -> Result<ApplyChunkResult, Error> + Send>>,
+    ) {
+        let sc = self.apply_chunks_sender.clone();
+        // TODO: uncomment this line when we are ready to move apply_chunks async
+        //rayon::spawn(move || {
+        let res = do_apply_chunks(work);
+        // we can safely unwrap here because error means the receiver is deallocated, which means
+        // the chain thread is shut down, and the node already crashed
+        sc.send((block_hash, res)).unwrap();
+        //});
+    }
+
+    /// Check if there are any block that finishes applying chunks. Run postprocessing on the block.
+    fn try_postprocess_block(
```

Comment:
Alternative name: *accept* processed blocks? Or maybe `postprocess_ready_blocks`? I don't like the `try_` suffix -- try means something might fail, not that something might be no-necessary.